api/pkg/service/auth/auth.go

```diff
@@ -0,0 +1,117 @@
+// Copyright © 2020 The Tekton Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/google/go-github/github"
+	"github.com/jinzhu/gorm"
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+
+	"github.com/tektoncd/hub/api/gen/auth"
+	"github.com/tektoncd/hub/api/pkg/app"
+	"github.com/tektoncd/hub/api/pkg/db/model"
+)
+
+type service struct {
+	logger        *zap.SugaredLogger
+	db            *gorm.DB
+	conf          *oauth2.Config
+	jwtSigningKey string
+	defaultScopes []string
+}
+
+var (
+	invalidCode   = auth.MakeInvalidCode(fmt.Errorf("invalid authorization code"))
+	internalError = auth.MakeInternalError(fmt.Errorf("failed to authenticate"))
+)
+
+// New returns the auth service implementation.
+func New(api app.Config) auth.Service {
+	return &service{api.Logger(), api.DB(), api.OauthConfig(), api.JWTSigningKey(), api.Data().Default.Scopes}
+}
+
+// Login authenticates users against GitHub Oauth
+func (s *service) Login(ctx context.Context, p *auth.LoginPayload) (res *auth.LoginResult, err error) {
+
+	// gets access_token for user using authorization_code
+	token, err := s.conf.Exchange(oauth2.NoContext, p.Code)
+	if err != nil {
+		return nil, invalidCode
+	}
+
+	// gets user details from github using the access_token
+	oauthClient := s.conf.Client(oauth2.NoContext, token)
+	ghClient := github.NewClient(oauthClient)
+	ghUser, _, err := ghClient.Users.Get(oauth2.NoContext, "")
+	if err != nil {
+		s.logger.Info(err)
+		return nil, internalError
+	}
+
+	// adds user in db if not exist
+	user, err := s.addUser(ghUser)
+	if err != nil {
+		return nil, err
+	}
+
+	// creates jwt using user details
+	jwt, err := s.createJWT(user)
+	if err != nil {
+		return nil, err
+	}
+
+	return &auth.LoginResult{Token: jwt}, nil
+}
+
+func (s *service) addUser(user *github.User) (*model.User, error) {
+
+	q := s.db.Model(&model.User{}).Where(&model.User{GithubLogin: user.GetLogin()})
+
+	newUser := &model.User{
+		GithubName:  user.GetName(),
+		GithubLogin: user.GetLogin(),
+	}
+	if err := q.FirstOrCreate(newUser).
+		Scan(newUser).
+		Error; err != nil {
+		s.logger.Error(err)
+		return nil, internalError
+	}
+
+	return newUser, nil
+}
+
+func (s *service) createJWT(user *model.User) (string, error) {
```

Comment:
can we make this an independent  function and not a function of the `service` ?
the `s.logger` can be done by the caller.

Reply:
I moved that generic code to a package, created this map here because the user data is specific to this service.
We can also do `jwtToken.createForNormalUser` which will call `create` in that package.
WDYT? also what should be package name?